test/e2e/util_gatewayapi_test.go

```diff
@@ -590,6 +623,103 @@ func assertGatewaySuccessful(t *testing.T, namespace, name string) (*gatewayapiv
 	return gw, nil
 }
 
+func waitForGatewayListenerCondition(t *testing.T, gatewayName types.NamespacedName, listenerName string, conditions ...metav1.Condition) error {
+	t.Helper()
+
+	gateway := &gatewayapiv1.Gateway{}
+	expected := gatewayListenerConditionMap(conditions...)
+	current := map[string]string{}
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, gatewayName, gateway); err != nil {
+			t.Logf("failed to get gateway %s: %v", gatewayName.Name, err)
+			return false, nil
+		}
+
+		var listenerStatus *gatewayapiv1.ListenerStatus
+		for _, ls := range gateway.Status.Listeners {
+			if string(ls.Name) == listenerName {
+				listenerStatus = &ls
+				current = gatewayListenerConditionMap(listenerStatus.Conditions...)
+				return conditionsMatchExpected(expected, current), nil
+			}
+		}
+
+		if listenerStatus == nil {
+			t.Logf("listener %s not found in gateway %s", listenerName, gatewayName.Name)
+			return false, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		t.Errorf("Expected conditions: %v\n Current conditions: %v", expected, current)
+		if gateway != nil {
+			if gwStatusPretty, err := json.MarshalIndent(gateway.Status, "", "  "); err != nil {
+				t.Fatal(err)
+			} else {
+				t.Logf("Gateway %s/%s status: %s\n", gateway.Namespace, gateway.Name, string(gwStatusPretty))
+			}
+		}
+	}
+	return err
+}
+
+func gatewayListenerConditionMap(conditions ...metav1.Condition) map[string]string {
+	conds := map[string]string{}
+	for _, cond := range conditions {
+		conds[string(cond.Type)] = string(cond.Status)
+	}
+	return conds
+}
+
+// assertListenerDNSRecord checks to make sure a DNSRecord exists in a ready state for the specified gateway listener,
+// and returns an error if not.
+func assertListenerDNSRecord(t *testing.T, gatewayName, listenerHostname string) error {
+	t.Helper()
+
+	if !strings.HasSuffix(listenerHostname, ".") {
+		listenerHostname += "."
+	}
+
+	var dnsRecord *v1.DNSRecord
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 3*time.Minute, false, func(ctx context.Context) (bool, error) {
+		dnsRecords := &v1.DNSRecordList{}
+		if err := kclient.List(ctx, dnsRecords, client.InNamespace(operatorcontroller.DefaultOperandNamespace)); err != nil {
+			t.Logf("failed to list DNSRecords: %v; retrying...", err)
+			return false, nil
+		}
+
+		for _, record := range dnsRecords.Items {
+			if record.Labels["gateway.networking.k8s.io/gateway-name"] == gatewayName &&
+				record.Spec.DNSName == listenerHostname {
```

Comment:
Doublecheck on whether you need to add a trailing `.`.

Reply:
To debug, you can just add a line that prints out the record labels and spec.dnsname.